### DISCUSSION TITLE
Generic method to count resources based on prisma model name

### DISCUSSION THREAD

**Discussion by isaacgr** [prisma/prisma#19599] [Q&A] [answered]
Not quite finding what I'm looking for. I'm wondering if there's a way to make a generic 'count' method on, what I'm hoping, is a generic model class.

```
export default class Model  {
	private readonly database: PrismaClient["partnumber" | "contributor"];

	constructor(database: PrismaClient["partnumber" | "contributor"]) {
		this.database = database;
	}

	count(): Promise<number> {
		return new Promise(async (resolve, reject) => {
			try {
				resolve(await this.database.count());
			} catch (e: any) {
				reject(new DatabaseError(e.message));
			}
		});
	}
}
```
I get the following error on 'count':

![image](https://github.com/prisma/prisma/assets/20640588/8035fca1-3953-4feb-916f-22a502ea142b)

I have several tables that I'd like to re-use the same calls on (count, get by id, get all) for a REST api. So I'd like to pass in the prisma client model name depending on which resource I'm accessing.

If there's a better way I'm open to that as well.

**Comment by nurul3101** [comment, marked as answer]
Hello @isaacgr 👋 

Here's a GitHub Discussion discussing the same use case

- https://github.com/prisma/prisma/discussions/18478

Can you have a look?

**Reply by nurul3101** [reply]
I am checking to see if you had a chance to look at my last response.

It would be great if you could mark this Discussion as answered to indicate that it has been resolved.
Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by isaacgr** [reply]
Hey sorry,

Let me look into this and see this week. Will definitely respond when I have some results. Thanks.

**Reply by nurul3101** [reply]
Sure, no worries!
Thanks for getting back and letting us know! 🙂

**Comment by isaacgr** [comment]
I get the following errors running what was pasted in the comment there. I'm still new to prisma so have to look into these separately but just an update.

```
    src/db/index.ts:15:17 - error TS2339: Property '$extends' does not exist on type 'PrismaClient<PrismaClientOptions, never, RejectOnNotFound | RejectPerOperation | undefined>'.

    15   this.database.$extends({
                       ~~~~~~~~
    src/db/index.ts:20:31 - error TS2694: Namespace '"index".Prisma' has no exported member 'Payload'.

    20       field: K & keyof Prisma.Payload<T, "findMany">["scalars"],
                                     ~~~~~~~
    src/db/index.ts:21:21 - error TS2694: Namespace '"index".Prisma' has no exported member 'Payload'.

    21       value: Prisma.Payload<T, "findMany">["scalars"][K]

    src/db/index.ts:23:26 - error TS2339: Property 'getExtensionContext' does not exist on type 'typeof Prisma'.                                                                                                                                                                                                                                          23       const ctx = Prisma.getExtensionContext(this) as any;   
```

**Reply by isaacgr** [reply]
Just some typing issues. Added `@ts-ignore` above those lines and it cleared those up.

Getting this now

```
Extensions are not yet generally available, please add `clientExtensions` to the `previewFeatures` field in the `generator` block in the `schema.prisma` file.
```

After adding 

```
previewFeatures = ["clientExtensions"]
```

to the schema.prisma file it looks like that doesnt work initially.

Getting ` TypeError: this.database.countSomething is not a function `.

I'll dig into that a bit more.

**Reply by isaacgr** [reply]
Reading the docs helps.

```
When you call a method in an extension, use the constant name from your $extends statement, not prisma. In the above example, xprisma.user.signUp works, but prisma.user.signUp does not, because the original prisma is not modified.
```
[Here](https://www.prisma.io/docs/concepts/components/prisma-client/client-extensions/model#:~:text=When%20you%20call%20a%20method%20in%20an%20extension%2C%20use%20the%20constant%20name%20from%20your%20%24extends%20statement%2C%20not%20prisma.%20In%20the%20above%20example%2C%20xprisma.user.signUp%20works%2C%20but%20prisma.user.signUp%20does%20not%2C%20because%20the%20original%20prisma%20is%20not%20modified.)

So yes. Looks like this is working after making that change.

**Comment by isaacgr** [comment]
Hey sorry just a question on this.

I notice that when I create an extended client its not a `PrismaClient` type, but a `GetClient` type.

I cant find anywhere directly referencing where this should be exported from however (to explicitly reference for function/variable types).

![image](https://github.com/prisma/prisma/assets/20640588/095dbdc3-abd6-4754-9a93-e25902dede22)

**Reply by isaacgr** [reply]
I found this article which helped. https://echobind.com/post/extending-types-for-prisma-extensions-in-nextjs

The main bit here
```js
const extendedPrismaClient = () => {
  const prisma = new PrismaClient({
    log: logOptions,
  });

  const extendedPrisma = prisma.$extends({
    model: {
      user: {
        async signUp(args: SignUpArgs): Promise<UserWithRelations> {
          return await prisma.user.create({
            ...args,
          });
        },
      },
    },
  });

  return extendedPrisma;
};

export type ExtendedPrismaClient = ReturnType<typeof extendedPrismaClient>;
```

**Reply by nurul3101** [reply]
Thanks for sharing! 🙌
